docs/examples/lipstick-pig.md

```diff
@@ -0,0 +1,315 @@
+
+flatten = lambda l: [item for sublist in l for item in sublist]
+
+def calc_axis(pair_list, language_model):
+    return [language_model[t1] - language_model[t2] for (t1, t2) in pair_list]
+
+def make_correlation_plot(pairs, language_model, metric="cosine"):
+    axes = [calc_axis(p, language_model) for p in pairs]
+    emb_pairs = EmbeddingSet(*flatten(axes))
+    emb_pairs.plot_distance(metric=metric)
+
+make_correlation_plot(pairs=all_pairs, language_model=lang_ft)
+```
+
+This code generates a similarity chart for fasttext embeddings, shown below.
+
+![](imgs/bias.png)
+
+Notice, that we indeed see correlation. The "gender" direction seems to correlate with the "doctor-nurse" direction as well. This does not bode well.
+
+## Projections
+
+We observe bias that we do not want to have. So it's natural to ask: can we remove it?
+
+There is a popular technique that proposes to filter out the "gender"-direction. If we can quantify the gender direction then we might also be able to project all the vectors in our set away from it. The 2D plot below demonstrates this idea.
+
+<details>
+  <summary><b>Plot Code</b></summary>
+```python
+from whatlies import Embedding
+
+man   = Embedding("man", [0.5, 0.1])
+woman = Embedding("woman", [0.5, 0.6])
+king  = Embedding("king", [0.7, 0.33])
+queen = Embedding("queen", [0.7, 0.9])
+
+man.plot(kind="arrow", color="blue")
+woman.plot(kind="arrow", color="red")
+king.plot(kind="arrow", color="blue")
+queen.plot(kind="arrow", color="red")
+
+(queen - king).plot(kind="arrow", color="pink", show_ops=True)
+(man | (queen - king)).plot(kind="arrow", color="pink", show_ops=True)
+plt.axis('off');
+```
+</details>
+
+![](imgs/logo.png)
+
+In this example you can see what might happen if we project $v_{man}$ away from the $v_{queen} - v_{king}$ axis we get a new vector $v_{man} | (v_{queen} - v_{king})$.
+
+You can even see that if in this 2D example;
+
+$$v_{man} | (v_{queen} - v_{king}) \approx v_{woman} | (v_{queen} - v_{king})$$
+
+This suggests that we can use linear algebra to "filter" away the gender information as well as the gender bias.
+
+## Post-Processing
+
+```python
+def make_debias_correlation_plot(pairs, language_model, metric='cosine'):
+    # Calculate the embeddings just like before.
+    axes = [calc_axis(p, language_model) for p in pairs]
+    emb_pairs = EmbeddingSet(*flatten(axes))
+
+    # Calculate the "gender"-direction
+    norm_emb = EmbeddingSet(
+        (language_model['man'] - language_model['woman']),
+        (language_model['king'] - language_model['queen']),
+        (language_model['father'] - language_model['mother'])
+    ).average()
+
+    # Project all embeddings away from this axis.
+    emb_pairs = emb_pairs | norm_emb
+
+    # Plot the result.
+    emb_pairs.plot_distance(metric=metric)
+
+make_debias_correlation_plot(pairs=all_pairs, language_model=lang_ft)
+```
+
+We'll now display the "before" as well as "after" chart.
+
+![](imgs/before-after.png)
```

Comment:
Something strange is happening here because the "before" plot colors is much brighter than its previous counter-part (i.e. `bias.png`), so much so that I can almost sense the bias in the "before" plot (unlike the `bias.png` plot)??!!!

Reply:
Whoa! Well spotted. The first chart was wrong!

Reply:
Fixed!